### PR TITLE
EQL: Sequences will now support nano-timestamps

### LINKED ISSUE #68812: EQL: Support date_nanos type for `@timestamp`
Currently the sequence matching assumes that the type of the `@timestamp` is `date`,
so only milliseconds precision is supported.

Should support for `date_nanos` be implemented as well?

### PR DESCRIPTION
**Pull request by bpintea**
With this change nanosecond-resolution timestamps can be used to drive
the EQL sequence query.

Closes #68812.

### CODE REVIEW

**Review by astefan — COMMENTED** [review]
Left few comments.

File: x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java [hunk 1]
```diff
@@ -175,14 +178,28 @@ private static void loadData(RestHighLevelClient client, String indexName, boole
...
         Long ts = (Long) object;
         // currently this is windows filetime
         entry.put("@timestamp", winFileTimeToUnix(ts));
     }
 
+    private static void timestampToUnixNanos(Map<String, Object> entry) {
+        Object object = entry.get("timestamp");
+        assertThat(object, instanceOf(Long.class));
+        // interpret the value as nanos since the unix epoch
+        String timestamp = entry.get("timestamp").toString();
```
[the remark this answers is not in the document]
Comment (astefan) [reply x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java:193]: I meant
```suggestion
        String timestamp = object.toString();
```

[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java:193]: Ah, you're right, thanks (_inside_ vs _assigned_ threw me off).
I'll fix this with any next round of changes.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java [hunk 2]
```diff
@@ -339,4 +340,27 @@ public String toString() {
                 completed.size(),
                 stageToKeys);
     }
+
+    private static long timestampDiffNanos(Object ts1, Object ts2) {
```
Comment (Luegg) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:344]: Maybe it would be good to have some dedicated unit tests for this method. Does it support diffing over the fully supported range of timestamps? Also, I can imagine the method is vulnerable to long overflow issues.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java [hunk 3]
```diff
@@ -339,4 +340,27 @@ public String toString() {
...
+        if (ts1 instanceof Instant) {
+            Instant i1 = (Instant) ts1;
+            Instant diff;
+            if (ts2 instanceof Instant) {
+                Instant i2 = (Instant) ts2;
+                diff = i1.minusNanos(i2.getNano()).minusSeconds(i2.getEpochSecond());
+            } else {
+                diff = i1.minusMillis(((Number) ts2).longValue());
+            }
+            timestampDiffNanos = diff.getEpochSecond() * 1_000_000_000L + diff.getNano();
```
Comment (Luegg) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:355]:
```suggestion
            return diff.getEpochSecond() * 1_000_000_000L + diff.getNano();
```

and the same for the other branches.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/search/OrdinalTests.java [hunk 4]
```diff
@@ -83,51 +85,74 @@ public void testCompareToSameTsSameTieNullDifferentImplicitTb() {
...
+        Ordinal after = new Ordinal(randomTimestampBetween(5000, 6000), (Comparable) randomLong(), randomLong());
 
         assertTrue(after.afterOrAt(before));
         assertTrue(after.afterOrAt(after));
         assertFalse(before.afterOrAt(after));
     }
+
+    static Object randomTimestamp() {
+        if (randomBoolean()) {
+            return randomLongBetween(-3155701416721920000L, 3155688986440319900L);
```
Comment (Luegg) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/search/OrdinalTests.java:138]: where do these bounds come from? Will timestamps never be outside of these ranges?

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Ordinal.java [hunk 5]
```diff
@@ -9,26 +9,33 @@
...
+
+import java.time.Instant;
 import java.util.Objects;
 
 public class Ordinal implements Comparable<Ordinal>, Accountable {
 
     private static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(Ordinal.class);
 
-    private final long timestamp;
+    private final Object timestamp;
```
Comment (Luegg) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Ordinal.java:20]: I'm probably missing quite some context, but why not always using instant for holding the ts?

File: x-pack/plugin/eql/qa/common/src/main/resources/test_queries_date_nanos.toml [hunk 6]
```diff
@@ -4,20 +4,20 @@
 [[queries]]
 name = "filterOnDateNanosWithMillis"
 query = '''
-    process where @timestamp == "1970-01-02T12:31:25.510Z"
+    process where @timestamp == "1974-03-02T19:53:16.510Z"
```
Comment (astefan) [inline x-pack/plugin/eql/qa/common/src/main/resources/test_queries_date_nanos.toml:7]: Why this change in test data? I got from you test changes that the data itself didn't actually change, I'm curious why have you changed these tests and, at the same time, `expected_event_ids` remained the same.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Ordinal.java [hunk 7]
```diff
@@ -120,7 +126,21 @@ public boolean afterOrAt(Ordinal other) {
...
+        long millisOfSecond = nanos / 1_000_000L;
+        return (instant.getEpochSecond() * 1000L + millisOfSecond) + "." + (nanos - millisOfSecond * 1_000_000);
+    }
+
+    private static int timestampCompare(Object ts1, Object ts2) {
+        return ts1 instanceof Instant
+            ? (ts2 instanceof Instant ? ((Instant) ts1).compareTo((Instant) ts2) :
+                ((Instant) ts1).compareTo(Instant.ofEpochMilli(((Number) ts2).longValue())))
+            : (ts2 instanceof Instant ? Instant.ofEpochMilli(((Number) ts1).longValue()).compareTo((Instant) ts2) :
+                Long.compare(((Number) ts1).longValue(), ((Number) ts2).longValue()));
```
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Ordinal.java:144]: Isn't the ternary operator being a bit overused here? Wouldn't an `if/else` be more easy to follow visually?

Also, looking at the changes in this PR, having `instanceof Instant` logic separate as a utility method would make more sense to show the difference between a date-with-nanos and date-with-millis?

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/assembler/ImplicitTiebreakerTests.java [hunk 8]
```diff
@@ -65,7 +65,7 @@ public void query(QueryRequest r, ActionListener<SearchResponse> l) {
                 int previous = ordinal - 1;
                 // except the first request, the rest should have the previous response's search_after _shard_doc value
                 assertArrayEquals("Elements at stage " + ordinal + " do not match",
-                    r.searchSource().searchAfter(), new Object[] { (long) previous, implicitTiebreakerValues.get(previous) });
+                    r.searchSource().searchAfter(), new Object[] { String.valueOf(previous), implicitTiebreakerValues.get(previous) });
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/assembler/ImplicitTiebreakerTests.java:68]: Leftover from undo-ing the implicit tiebreaker changes?

**Review by astefan — APPROVED** [review]
LGTM in general. I've left some questions.

**Review by costin — CHANGES_REQUESTED** [review]
The Instant/Long handling can be better encapsulated. I'm +1 in not having a wrapper class (see my other comment) around the Long/Instant however as it stands, the code that handles the differences is spread across multiple classes making it hard to follow.
Instead I propose to consolidate it in a single class (potentially in the ordinal package) and move there all the methods that perform comparison, difference, toString() etc...
That should simplify some things since constants like 1_000_000_000 are defined once and better documented while calling classes can simply call:
`timestamps.diff(ordinal.timestamp, newOrdinal.timestamp)`.

As a separate note, it's worth checking whether we can avoid some of the boxing/autoboxing happening by using `Object`. Previously `long` was used however due to `Instant` now `Object` is passed around as a signature which means anything that happens with `long` creates objects - so maybe adding a wrapping class (Timestamp) with different implementations, on for `long` and one for `Instant` might work out after all.

**Comment by costin** [comment]
I'm concerned that the tests have to be changed to make this change work out - that should not be the case.

File: x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java [hunk 9]
```diff
@@ -175,14 +178,28 @@ private static void loadData(RestHighLevelClient client, String indexName, boole
...
+        assertThat(timestamp.length(), greaterThan(12));
+        // avoid double approximations and BigDecimal ops
+        String millis = timestamp.substring(0, 12);
+        String milliFraction = timestamp.substring(12);
+        // strip the fractions right away if not actually present
+        entry.put("@timestamp", milliFraction.equals("000000") ? millis : millis + "." + milliFraction);
+        entry.put("timestamp", ((long) object)/1_000_000L);
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java:201]: Just like for the existing millis tests, `timestamp` is populated but not used in the qa tests. Having it with a meaningful value allows however quick millis vs nanos comparison tests though (used mostly for troubleshooting, only changing the `timestamp_field`).

File: x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java [hunk 10]
```diff
@@ -86,33 +89,33 @@ public static void loadDatasetIntoEs(RestHighLevelClient client,
...
         //
         // Aux Index
         //
-        load(client, TEST_EXTRA_INDEX, false, p);
+        load(client, TEST_EXTRA_INDEX, null, null, p);
         //
         // Date_Nanos index
         //
-        // The data for this index are identical to the endgame-140.data with only the values for @timestamp changed.
+        // The data for this index is loaded from the same endgame-140.data sample, only having the mapping for @timestamp changed.
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java]: Good point, thanks, added.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java (hunk 2 above)
Comment (bpintea) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java]: This has been now refactored and mostly covered by OrdinalTests.java.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java (hunk 3 above)
Comment (bpintea) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java]: Refactored.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/search/OrdinalTests.java (hunk 4 above)
Comment (bpintea) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/search/OrdinalTests.java]: Those came from Instant, that has some (private) ranges for seconds. But this has been now improved.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Ordinal.java (hunk 5 above)
Comment (bpintea) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Ordinal.java]: Mostly for performance reasons, since timestamp handling is a core sequences operation and presumably most queries run on millis (as it had so far), i.e. longs.

File: x-pack/plugin/eql/qa/common/src/main/resources/test_queries_date_nanos.toml (hunk 6 above)
Comment (bpintea) [reply x-pack/plugin/eql/qa/common/src/main/resources/test_queries_date_nanos.toml:7]: > I got from you test changes that the data itself didn't actually change

Yes, it did change: the nano tests used before a copy of `endgame-140.data` that was modified to match the few tests contained in `test_queries_date_nanos.toml`. With this PR `eql_date_nanos.data` has been removed and tests setup to use with the same data - `endgame-140.data` - [with different interpretation of the timestamp value](https://github.com/elastic/elasticsearch/pull/76953#discussion_r696092631). This allows to run the same tests - `test_queries.toml` on both millis and nanos. But this also requires the event nano tests - `test_queries_date_nanos.toml`  - changed in order to match the existing results (that are kept unchanged).

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Ordinal.java (hunk 7 above)
Comment (bpintea) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Ordinal.java]: This has been refactored.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/assembler/ImplicitTiebreakerTests.java (hunk 8 above)
Comment (bpintea) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/assembler/ImplicitTiebreakerTests.java:68]: This has be a string, since otherwise the value - before a Long/Instant, now a Timestamp - can't be used when processing downstream the search object.

File: x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java [hunk 11]
```diff
@@ -44,8 +46,9 @@
  * Loads EQL dataset into ES.
  *
  * Checks for predefined indices:
- * - endgame-140 - for existing data
- * - extra       - additional data
+ * - endgame-140    - for existing data
+ * - eql_date_nanos - same as endgame-140, but with nano-precision timestamps
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/DataLoader.java]: Applied.

File: x-pack/plugin/eql/qa/common/src/main/resources/test_queries.toml [hunk 12]
```diff
@@ -861,7 +861,7 @@ expected_event_ids  = [67, 68, 69, 70,
 [[queries]]
 name = "sequencesOnDifferentEventTypes7"
 query = '''
-sequence with maxspan=500ms
+sequence with maxspan=50ms
```
[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/eql/qa/common/src/main/resources/test_queries.toml:865]: I've added a comment to the test and [explained further](https://github.com/elastic/elasticsearch/pull/76953#discussion_r699568002) the change intention, to allow the same sequence queries be executed on both millis and nanos timestamps.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Timestamp.java [hunk 13]
```diff
@@ -42,14 +42,9 @@ public static Timestamp of(String milliseconds) {
         return timestamp;
     }
 
-    // time delta in nanos from other to instance
+    // time delta in nanos between this and other instance
     public long diff(Timestamp other) {
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Timestamp.java:46]: This could be improved by `NANOS.between(this, other)` which behind the scenes does the math computation in this method.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/MillisTimestamp.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    private final long timestamp;
+    private Instant instant = null;
+
+    MillisTimestamp(long millis) {
+        timestamp = millis;
+    }
+
+    @Override
+    public int compareTo(Timestamp other) {
+        return other instanceof MillisTimestamp ? Long.compare(timestamp, ((MillisTimestamp) other).timestamp) : -other.compareTo(this);
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/MillisTimestamp.java:23]: to avoid the surprising `- ` you could just delegate to the parent - `super.compareTo(this)`.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/MillisTimestamp.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Override
+    public long diff(Timestamp other) {
+        return other instanceof MillisTimestamp ? (timestamp - ((MillisTimestamp) other).timestamp) * NANOS_PER_MILLI : -other.diff(this);
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/MillisTimestamp.java:28]: See above.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Timestamp.java (hunk 13 above)
Comment (costin) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Timestamp.java:46]: Also delta, until or between sound better than `diff`.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/assembler/SequenceSpecTests.java [hunk 16]
```diff
@@ -85,8 +86,8 @@ public String hitName() {
         static final TimestampExtractor INSTANCE = new TimestampExtractor();
 
         @Override
-        public Long extract(SearchHit hit) {
-            return (long) hit.docId();
+        public Object extract(SearchHit hit) {
```
Comment (costin) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/assembler/SequenceSpecTests.java:89]: The return time should be Timestamp not Object.

**Review by costin — APPROVED** [review]
Left some comments around cosmetics - LGTM.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/assembler/SequenceSpecTests.java (hunk 16 above)
Comment (bpintea) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/assembler/SequenceSpecTests.java]: Fixed, thanks.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/MillisTimestamp.java (hunk 15 above)
Comment (bpintea) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/MillisTimestamp.java]: Ditto.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/MillisTimestamp.java (hunk 14 above)
Comment (bpintea) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/MillisTimestamp.java]: Reverted to previous format.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Timestamp.java (hunk 13 above)
Comment (bpintea) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/Timestamp.java]: Went for delta. (`until` suggests that `other` is expected to be past `this`, which is not and `between` might require a range argument)
